### PR TITLE
Replace Blake2b512 with blake2b_simd to gain 10% performance improvement

### PR DESCRIPTION
**Pull request by alexzhenyu**
## Motivation

Replace Blake2b512 with blake2b_simd to gain 10% performance improvement

## Test Plan

cli：cargo test test_coinbase_puzzle --release

## Related PRs

(Link your related PRs here)

### CODE REVIEW

**Comment by Pratyush** [comment]
Hi @alexzhenyu!

Thanks for the PR! Do you mind running `cargo +nightly fmt` in the repo root to fix the formatting issues?

**Comment by alexzhenyu** [comment]
> Hi @alexzhenyu!
> 
> Thanks for the PR! Do you mind running `cargo +nightly fmt` in the repo root to fix the formatting issues?

Sure. Thanks for reminding me.

**Review by Pratyush — DISMISSED** [review]
Thanks, this is great! @howardwu this is a good speedup; the only downside is the added dependency.

Also @alexzhenyu, does the behaviour of the new hash match that of the old one?

File: synthesizer/src/coinbase_puzzle/hash.rs [hunk 1]
```diff
@@ -53,7 +56,11 @@ pub fn hash_commitment<E: PairingEngine>(commitment: &KZGCommitment<E>) -> Resul
     ensure!(bytes.len() == 96, "Invalid commitment byte length for hashing");
 
     // Return the hash of the commitment.
-    Ok(E::Fr::from_bytes_le_mod_order(&blake2::Blake2b512::digest(&bytes)))
+    //Ok(E::Fr::from_bytes_le_mod_order(&blake2::Blake2b512::digest(&bytes)))
+
```
Comment (Pratyush) [inline synthesizer/src/coinbase_puzzle/hash.rs]:
```suggestion
```

**Comment by marco-storswift** [comment]
@Pratyush  we test the  blake2b_simd opt in hash_to_coefficients function,
        let a0 = Local::now().timestamp_nanos();
        let blake2b512: Vec<_> = cfg_into_iter!(0..num_coefficients)
            .map(|counter| {
                let mut input_with_counter = [0u8; 36];
                input_with_counter[..32].copy_from_slice(&hash);
                input_with_counter[32..].copy_from_slice(&counter.to_le_bytes());
                F::from_bytes_le_mod_order(&blake2::Blake2b512::digest(input_with_counter))
            })
            .collect();
        let a1 = Local::now().timestamp_nanos();
        let blake2b_simd: Vec<_> = cfg_into_iter!(0..num_coefficients)
            .map(|counter| {
                let mut input_with_counter = [0u8; 36];
                input_with_counter[..32].copy_from_slice(&hash);
                input_with_counter[32..].copy_from_slice(&counter.to_le_bytes());
                let result = Params::new()
                    .hash_length(64)
                    .hash(&input_with_counter);
                let input_hash= result.as_bytes();
                F::from_bytes_le_mod_order(input_hash)
            })
            .collect();
        let a2 = Local::now().timestamp_nanos();
        assert_eq!(blake2b512, blake2b_simd);
        eprintln!("{}  hash_to_poly blake2b_simd use {} ",  Local::now().to_rfc3339(), a2 - a1);
        eprintln!("{}  hash_to_poly Blake2b512 use {} ",  Local::now().to_rfc3339(), a1 - a0);
        got the result 
        2022-11-09T10:24:17.200472400+08:00  hash_to_poly blake2b_simd use 4192700 
2022-11-09T10:24:17.200490200+08:00  hash_to_poly Blake2b512 use 4777700 
2022-11-09T10:24:17.341930100+08:00  hash_to_poly blake2b_simd use 3993700 
2022-11-09T10:24:17.341953900+08:00  hash_to_poly Blake2b512 use 4719900 
2022-11-09T10:24:17.493659100+08:00  hash_to_poly blake2b_simd use 3983700 
2022-11-09T10:24:17.493683100+08:00  hash_to_poly Blake2b512 use 4785800 
2022-11-09T10:24:17.631972800+08:00  hash_to_poly blake2b_simd use 4784900 
2022-11-09T10:24:17.632001500+08:00  hash_to_poly Blake2b512 use 5432200

**Comment by alexzhenyu** [comment]
> Thanks, this is great! @howardwu this is a good speedup; the only downside is the added dependency.
> 
> Also @alexzhenyu, does the behaviour of the new hash match that of the old one?

Yes, we verified the behavior by running test_coinbase_puzzle. Now our provers submitted the puzzle solutions computed  by blake2b_simd method successfully, and the solution can be found on the site( https://www.aleo123.io/block/ab1jxra2g7vngn7lfmjm6ts9w4xdy9w4frnrnuuhajk3upfuwh2eg9sgs5ddw)

**Review by howardwu — APPROVED** [review]
LGTM, thanks for the PR @alexzhenyu 🙏

**Comment by alexzhenyu** [comment]
> LGTM, thanks for the PR @alexzhenyu 🙏

Aleo is  GOAT！

**Comment by raychu86** [comment]
Thanks for the PR! Is this change WASM compatible?

**Comment by alexzhenyu** [comment]
> compatible

It's a great question. I have no idea how to identify such WASM-compatible stuff. Would you like to help us?

**Comment by howardwu** [comment]
I just did a quick check and confirmed `blake2b_simd` compiles in WASM. Thanks for bringing this up @raychu86

**Comment by howardwu** [comment]
@alexzhenyu I tried to reproduce the performance improvements and was unable to.

Could you try running this and let me know your findings?
```
git checkout feat/blake2b-tests
cd synthesizer
cargo test simd -- --nocapture
```

**Comment by Pratyush** [comment]
@howardwu Are you testing on ARM machines? I presume the speed up applies primarily to x86_64 machines; I’m guessing that the underlying crate does not have SIMD ops for arm64.

**Comment by howardwu** [comment]
Yeah, I had tested it on a Ubuntu `x86`. I confirmed SSE was enabled

**Comment by alexzhenyu** [comment]
> @alexzhenyu I tried to reproduce the performance improvements and was unable to.
> 
> Could you try running this and let me know your findings?
> 
> ```
> git checkout feat/blake2b-tests
> cd synthesizer
> cargo test simd -- --nocapture
> ```

We are sorry that we misled you. Our previous test did not sample enough data, which led to the wrong conclusion. I have done tests on both AMD & lntel machines and observed that SIMD does not beat Hash2b-512. We will continue to work on the acceleration of the hash implementation.

**Comment by howardwu** [comment]
No worries and thank you for taking the time to improve the puzzle and double check this work!

**Pull request closed without merging**